### PR TITLE
Move Frustum, HalfSpace from bevy_render to bevy_math

### LINKED ISSUE #13878: `bevy_render::Frustrum` should live in `bevy_math`
## What problem does this solve or what need does it fill?

Our existing [`Frustrum`](https://dev-docs.bevyengine.org/bevy/render/primitives/struct.Frustum.html) type is used for computing what cameras should see.

However, as this existed before the addition of our mathematical primitive types, it a) doesn't live in `bevy_math` like the others and b) is missing all of the standard expected functionality for it.

## What solution would you like?

1. Move `HalfSpace` into `bevy_math` unchanged.
2. Move `Frustrum` into `bevy_math` unchanged.
3. Implement gizmos for frustrums.

1 and 2 should be done in a single PR. 3 should be done in a follow-up.

We can later add all of the other weird and wonderful things (random sampling, meshability) to these types if people think it's worthwhile, but those steps are the key things.

Once things are moved, we can look into changing naming, representation and so on. Those should be done in tightly scoped PRs, with aggressive benchmarking for frustrum culling performance. This is directly in the rendering hot loop: changes here have a high impact, for both better and worse.

## Additional context

Raised [on Discord](https://discord.com/channels/691052431525675048/1203087353850364004/1251937356710281349). @superdump gave us rendering's blessing, as long as we don't regress performance.

The remaining [bevy_render::primitives](https://dev-docs.bevyengine.org/bevy/render/primitives/index.html) types should be moved / consolidated eventually, but gizmos for frustrums is the motivating requirement here.

### PR DESCRIPTION
**Pull request by torsteingrindvik**
Also move necessary baggage.

# Objective

Fixes #13878 

## Solution

- Move `Frustum` and `HalfSpace` from `bevy_render` to `bevy_math`
- Move the `Sphere` and `Aabb` structs from `bevy_render` to a legacy module in `bevy_math`, to be refactored later

## Changelog

- Move `Frustum` and `HalfSpace` from `bevy_render` to `bevy_math`

## Migration Guide

- `Frustum` and `HalfSpace` must now be imported from `bevy_math::primitives::{Frustum, HalfSpace}` instead of from `bevy_render`.

### CODE REVIEW

**Comment by torsteingrindvik** [comment]
When moving baggage into the legacy module (as suggested in https://github.com/bevyengine/bevy/issues/13878#issuecomment-2171800565) I had to also make docs for those, but since it's a legacy module I just mocked the docs. Let me know what should be done there (if not that).

Also the `bevy_render/src/lib.rs` registering of types seem kinda weird :thinking:

File: crates/bevy_math/src/primitives/mod.rs [hunk 1]
```diff
@@ -6,6 +6,13 @@ mod dim2;
 pub use dim2::*;
 mod dim3;
 pub use dim3::*;
+mod frustum;
+pub use frustum::Frustum;
+mod half_space;
+pub use half_space::HalfSpace;
+/// Clean me up!
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/primitives/mod.rs:13]: This needs real docs explaining why this exists, and linking to an issue made to track their clean-up.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Good, just a few notes:

1. IMO we should remove the legacy notes docs and simply blanket allow missing docs on the legacy_bevy_render module.
2. We should also remove `Sphere` from `bevy_render` during this PR rather than duplicating it, and note that in the migration guide.
3. We should make an issue specifically for eliminating `bevy_render::primitives` and link it in the module docs of both the `bevy_math` and `bevy_render` modules.

**Comment by torsteingrindvik** [comment]
> We should also remove `Sphere` from `bevy_render` during this PR rather than duplicating it, and note that in the migration guide.

Perhaps I misunderstood the goal/scope of point 1. and 2. in the issue, but could you elaborate?
`Sphere` _is_ deleted from `bevy_render` [here](https://github.com/bevyengine/bevy/pull/13882/files#diff-bfac89f1dcc3559a931d8c710dea14ac10a28ccad3f73361a5e271d9fe3eef68L112-L116).

Linking to changes seems a bit buggy, but it's deleted from `bevy_render/src/primitives/mod.rs`.

If you mean it's duplicated because it's now in the legacy module and we should remove that definition altogether then I would need to figure out the implications since the `Sphere` definition from `bevy_render` has a `center` field and the one in `bevy_math` doesn't.
I was probably missing some context from the conversation on Discord- I'm feeling a bit zapped for energy currently so depending on the scope I might put this up for adaption or close. But if I can get a clear picture it might be ok.

**Comment by alice-i-cecile** [comment]
> Sphere is deleted from bevy_render [here](https://github.com/bevyengine/bevy/pull/13882/files#diff-bfac89f1dcc3559a931d8c710dea14ac10a28ccad3f73361a5e271d9fe3eef68L112-L116).

I just misread the diff, sorry for the confusion!

**Review by BD103 — APPROVED** [review]

**Comment by torsteingrindvik** [comment]
> > Sphere is deleted from bevy_render [here](https://github.com/bevyengine/bevy/pull/13882/files#diff-bfac89f1dcc3559a931d8c710dea14ac10a28ccad3f73361a5e271d9fe3eef68L112-L116).
> 
> I just misread the diff, sorry for the confusion! I can make the follow-up issue, so the only thing I want from you is the missing_docs change.

Was questioning my sanity for a bit there :sweat_smile: 

I made a placeholder issue and linked to it in the PR: https://github.com/bevyengine/bevy/issues/13931 , will you edit it when needed?

**Review by alice-i-cecile — APPROVED** [review]
Great: I'll merge this as soon as 0.14 is out the door. Merging refactors like this will make cherrypicking fixes harder, so I'm going to hold off for just a bit.

**Comment by Jondolf** [comment]
> If you mean it's duplicated because it's now in the legacy module and we should remove that definition altogether then I would need to figure out the implications since the Sphere definition from bevy_render has a center field and the one in bevy_math doesn't.

`Sphere` is functionally like a bounding sphere, and its equivalent in `bevy_math` is the aptly named `BoundingSphere`. The two types have basically identical representations, except `BoundingSphere` has a bunch more functionality. Afaik the only missing method is `intersects_obb`.

--- 

As for the `Aabb`, we can't just remove it from `bevy_render` or deprecate it without a replacement since it's an actual component used during frustum culling. It should either (1) be a wrapper over `Aabb3d` living in `bevy_render`, or (2) remain roughly the same, but probably renamed to have a more specific name like `CullingAabb` or just `RenderAabb`.

The main "issue" with option (1) is that `Aabb` has a center + half-extents representation while `Aabb3d` uses min + max (which is *usually* better), so we need benchmarks if we intend to change it to use `Aabb3d`. I guess this might be fine to leave to a follow-up though, as long as it's not merged for 0.14.

File: crates/bevy_math/src/primitives/half_space.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+)]
+#[cfg_attr(
+    all(feature = "serialize", feature = "bevy_reflect"),
+    reflect(Serialize, Deserialize)
+)]
+#[cfg_attr(
+    all(feature = "bevy_ecs", feature = "bevy_reflect"),
+    reflect(Component)
+)]
+pub struct HalfSpace {
```
Comment (Jondolf) [inline crates/bevy_math/src/primitives/half_space.rs:51]: This should not be a component. The old one isn't either.

**Review by Jondolf — COMMENTED** [review]
Personally, I don't really like adding a legacy rendering module like this to `bevy_math` *especially* since it's not gated behind feature flags, and I don't really see the utility. It just feels like extra complexity and is not something `bevy_math` should have at all, even momentarily imo.

`Sphere` is a type that we should easily be able to just remove, and `Aabb` is a type we need to keep around in some form in `bevy_render` anyway. I'd just add an `aabb` module (or similar) in `bevy_render`, or even leave it where it was, and leave refactoring/renaming the type to a follow-up. Similarly for `Frustum`, if it needs to be a component, it just means that we need a newtype in `bevy_render`.

**Comment by alice-i-cecile** [comment]
@Jondolf how do you think we should tackle this broader initiative of "don't have weird primitives in bevy_render" as a whole? I really value your judgement here, and it's not clear what can / should be done first.

**Comment by Jondolf** [comment]
My response became so long that I just wrote it up into its own issue 😅 See #13945 for the full context.

TLDR concerning this PR: The `Aabb` and `Sphere` changes aren't as trivial and shouldn't be done here imo, and I think that adding the legacy module and `bevy_ecs` dependency in `bevy_math` just isn't a great idea even if they're temporary, *especially* given that we don't seem to have a clear plan of *how* we'd move forward with it yet.

My opinion is that this PR should just do what the PR title says:

1. Move `HalfSpace` into `bevy_math`. (done)
2. (Optional) Move `Frustum` into `bevy_math`. But:
	- Rename it to something like `ViewFrustum` to be more accurate and avoid overlap with e.g. `ConicalFrustum`.
	- Add a wrapper component in `bevy_render`. Not sure what name fits best since both lights and cameras can have the frustum component. Rendering people might be able to give input here. Either way, I don't really want components in `bevy_math`, even behind feature flags.
	- Add `intersects_obb` as a method on the wrapper component in `bevy_render` for now. It'd be nice to have a similar method in `bevy_math`, but again I don't really like adding a duplicate `Aabb` or a dependency on `bevy_render` there. And this is just simpler imo :)

No legacy modules or other unrelated changes, just these two (or just (1)). The `Aabb` and `Sphere` refactor can be done in a separate PR imo.

Feel free to disagree, these are just my thoughts on this :)

**Comment by torsteingrindvik** [comment]
> My response became so long that I just wrote it up into its own issue 😅 See #13945 for the full context.
> 
> TLDR concerning this PR: The `Aabb` and `Sphere` changes aren't as trivial and shouldn't be done here imo, and I think that adding the legacy module and `bevy_ecs` dependency in `bevy_math` just isn't a great idea even if they're temporary, _especially_ given that we don't seem to have a clear plan of _how_ we'd move forward afterwards.
> 
> My opinion is that this PR should just do what the PR title says:
> 
>     1. Move `HalfSpace` into `bevy_math`. (done)
> 
>     2. (Optional) Move `Frustum` into `bevy_math`. But:
>        
>        * Rename it to something like `ViewFrustum` to be more accurate and avoid overlap with e.g. `ConicalFrustum`.
>        * Add a wrapper component in `bevy_render`. Not sure what name fits best since both lights and cameras can have the frustum component. Rendering people might be able to give input here. Either way, I don't really want components in `bevy_math`, even behind feature flags.
>        * Add `intersects_obb` as a method on the wrapper component in `bevy_render` for now. It'd be nice to have a similar method in `bevy_math`, but again I don't really like adding a duplicate `Aabb` or a dependency on `bevy_render` there. And this is just simpler imo :)
> 
> 
> No legacy modules or other unrelated changes, just these two (or just (1)). The `Aabb` and `Sphere` refactor can be done in a separate PR imo.
> 
> Feel free to disagree, these are just my thoughts on this :)

Thanks for the explanation. This PR was originally just to address a chore when I had some extra time but there are more moving parts involved- so I think it's simplest for me to close this out for now.
The issue you created I think will serve the goals better.
